HIPCMS-901, Reviews for created content

thielep opened the pull request:
Added the option to create reviews for exhibits, routes and exhibit pages.

Commit:
* Added option to add reviews to exhibits and routes.

* Added option to get and change reviews of exhibits and routes.
Reviews can be approved and comments can be added to them.

Commit:
* Added review feature for exhibit pages.

* Removed unnecessary code

Commit:
* Added option to approve reviews through comments
One admin or if permitted the defined number of students are able to approve a review

* Updated permissions to get, create and change reviews

Commit:
Merge branch 'develop' into iss-hipcms-901-reviews

Commit:
Merge branch 'develop' into iss-hipcms-901-reviews

Commit:
Merge branch 'develop' into iss-hipcms-901-reviews

Commit:
Merged branch 'develop'

Commit:
- Updated reviews to work with previous update semantic changes.

Commit:
* Fix Resharper warnings

Commit:
* Fix resharper warnings

Commit:
* Fix resharper warnings

Commit:
* Fix resharper warnings

Commit:
Merge branch 'develop' into iss-hipcms-901-reviews

Commit:
* Added usage of databse.
Reviews are now saved in mongodb.

Commit:
* Resolved merge conflicts

Commit:
* Fix resharper warnings

Commit:
* Fix resharper warnings

mb236 committed:
Merge branch 'develop' into iss-hipcms-901-reviews

mb236 commented on HiP-DataStore/Controllers/ExhibitPagesController.cs:
```diff
@@ -280,6 +283,192 @@ public IActionResult GetReferenceInfo(int id)
+        [ProducesResponseType(404)]
+        public async Task<IActionResult> PostReviewAsync(int id, ReviewArgs args)
+        {
+            if (!ModelState.IsValid)
```
This code appears in all controllers. It looks like we could outsource this code and just pass the ResourceType and the id of the entity to it.

mb236 commented on HiP-DataStore/Controllers/ExhibitPagesController.cs:
```diff
@@ -280,6 +283,192 @@ public IActionResult GetReferenceInfo(int id)
+            if (!_reviewIndex.Exists(ResourceTypes.ExhibitPage.Name, id))
+                return NotFound(ErrorMessages.ReviewNotFound(ResourceTypes.ExhibitPage, id));
+
+            var reviewId = _reviewIndex.GetReviewId(ResourceTypes.ExhibitPage.Name, id);
```
Same for here

mb236 commented on HiP-DataStore.Model/Rest/ReviewArgs.cs:
```diff
@@ -0,0 +1,19 @@
+    public class ReviewArgs
+    {
+
+        public string Description { get; set; }
```
We should be able to merge ReviewArgs and ReviewUpdateArgs by using default values.

mb236 commented on HiP-DataStore.Model/Rest/ReviewUpdateArgs.cs:
```diff
@@ -0,0 +1,33 @@
+        public int EntityId { get; set; }
+
+        [JsonIgnore]
+        public List<Comment> Comments { get; set; } = new List<Comment>();
```
It might be more feasible to create a separate resource type for ReviewComments. With the current situation you would have to pass the whole Review object every time a comment was added. So I suggest we change this to a list of ints which are then the ids for the comments

mb236 commented on HiP-DataStore/Controllers/ExhibitPagesController.cs:
```diff
@@ -280,6 +283,192 @@ public IActionResult GetReferenceInfo(int id)
+        [ProducesResponseType(400)]
+        [ProducesResponseType(403)]
+        [ProducesResponseType(404)]
+        public IActionResult GetRreview(int id)
```
Typo

mb236 commented on HiP-DataStore/Controllers/ExhibitPagesController.cs:
```diff
@@ -280,6 +283,192 @@ public IActionResult GetReferenceInfo(int id)
+                return NotFound(ErrorMessages.ReviewNotFound(ResourceTypes.ExhibitPage, id));
+
+            var reviewId = _reviewIndex.GetReviewId(ResourceTypes.ExhibitPage.Name, id);
+            var review = _db.Get<Review>((ResourceTypes.Review, reviewId));
```
You should get the object from the event stream (by using the extension method on the event stream) instead of the database

Commit:
* Added ReviewComment Controller

* Outsourced same code from Exhibits, Routes and ExhibitsPages Controller to ReviewHelper

* Removed ReviewUpdateArgs, now only ReviewArgs are used

* Implemented minor requested changes

Commit:
* Fixed an error

Commit:
Merge branch 'iss-hipcms-901-reviews' of https://github.com/HiP-App/HiP-DataStore into iss-hipcms-901-reviews

Commit:
Merge branch 'develop' into iss-hipcms-901-reviews

mb236 commented on HiP-DataStore/Controllers/ReviewCommentsController.cs:
```diff
@@ -0,0 +1,130 @@
+            if (args.Approved)
+                newReviewArgs.Approved = ReviewHelper.IsReviewApproved(newReviewArgs.Comments, args.Approved, newReviewArgs.StudentsToApprove, User.Identity, _reviewCommentIndex);
+
+            await EntityManager.CreateEntityAsync(_eventStore, reviewComment, ResourceTypes.ReviewComment, reviewCommentId, User.Identity.GetUserIdentity());
```
Something looks wrong to me here. In the ResourceTypes class you define that the ReviewCommentArgs should be the corresponding class for the ReviewComment resource type. But here you pass an ReviewComment object instead of a ReviewCommentArgs object

mb236 commented on HiP-DataStore.Model/Entity/ReviewComment.cs:
```diff
@@ -0,0 +1,35 @@
+
+        public bool Approved { get; set; }
+
+        public int ReviewId { get; set; }
```
Do we need the connection from Reviews to ReviewComments in both directions? The Reviews already have list of the id's of the ReviewComments, this should be enough to model everything

mb236 commented on HiP-DataStore/Controllers/ExhibitsController.cs:
```diff
@@ -645,23 +642,21 @@ public IActionResult GetRreview(int id)
 
-            var reviewId = _reviewIndex.NextId(ResourceTypes.Exhibit);
-
             await EntityManager.CreateEntityAsync(_eventStore, newReview, ResourceTypes.Review, reviewId, User.Identity.GetUserIdentity());
```
Same here as for ReviewComments

mb236 commented on HiP-DataStore.Model/Entity/Review.cs:
```diff
@@ -48,10 +39,10 @@ public ReviewUpdateArgs CreateReviewUpdateArgs()
         public string EntityType { get; set; }
 
-        // Id of the entity the review belongs to
+        // ID of the entity the review belongs to
```
Currently there is a problem with int properties that no PropertyChanged event will be created for id 0. This is exactly what my [PR](https://github.com/HiP-App/HiP-DataStore/pull/124) adresses, so my PR should probably get merged first

Commit:
* Applied changes requested in review
** Removed reviewId from ReviewComments
** ReviewCommentArgs and ReviewArgs are now used instead of ReviewComment and Review, to create the entitiy
** fixed typo

Commit:
Merge branch 'develop' into iss-hipcms-901-reviews

Commit:
* Fixed Resharper warnings

mb236 commented on HiP-DataStore.Model/Rest/ReviewArgs.cs:
```diff
@@ -0,0 +1,19 @@
+
+        public string Description { get; set; }
+
+        public List<string> Reviewers { get; set; }
```
It should be made clear that a list of Id's has to be passed here and not a list of usernames. A comment should be enough

mb236 reviewed (changes requested):
Looks good, only one place where documentation should be added

Commit:
* Added a comment

Commit:
Merge branch 'develop' into iss-hipcms-901-reviews

mb236 committed:
Merge branch 'develop' into iss-hipcms-901-reviews

mb236 commented on HiP-DataStore/Controllers/ReviewCommentsController.cs:
```diff
@@ -0,0 +1,122 @@
+
+            var reviewCommentArgs = await _eventStore.EventStream.GetCurrentEntityAsync<ReviewCommentArgs>(ResourceTypes.ReviewComment, id);
+
+            var args = new ReviewCommentArgs()
```
Why not just reviewCommentArgs.Text = text?

thielep replied:
Maybe I don't get what you mean by that, but which objects would UpdateEntity compare then, if the old reviewCommentArgs are changed?

mb236 replied:
I forgot about that, than it's fine

mb236 commented on HiP-DataStore/Controllers/ReviewHelper.cs:
```diff
@@ -0,0 +1,106 @@
+                args.StudentsToApprove = 0;
+
+            // keep old values if no new ones are specified
+            if (args.Description == null)
```
I think this in inconsistent to our current behavior. By our current API design (e.g. for Exhibits) we assume that always the whole object is provided (including old values). This means that the copying of the old values is unnecessary.

mb236 commented on HiP-DataStore.Model/Rest/ReviewArgs.cs:
```diff
@@ -0,0 +1,42 @@
+        [JsonIgnore]
+        public int EntityId { get; set; }
+
+        public ReviewArgs CloneArgs()
```
I would prefer a constructor that accepts the another ReviewArgs object and copies the values from it.  But you also need an empty constructor as well (for the EventStoreLib)

Commit:
Merge branch 'develop' into iss-hipcms-901-reviews

Commit:
* Removed unnecessary value saving from ReviewHelper

* Added constructor to ReviewArgs

Commit:
* Fix Resharper warnings

Commit:
* Updated SDK versions
